README.md
# @mconnect/mcmail

- Send Email / Messages Package


HISTORY.md
# Historical logs

## Version 0.4.0

- Release date: May 18, 2022
- updated-deps-and-refactored-functions

## Version 0.3.1

- Release date: June 06, 2021
- updated mcresponse to v0.4.1

## Version 0.3.0

- Release date: Mar 08, 2020
- Refactored send-method, test, build and publish

## Version 0.2.0

- Release date: Dec 20, 2020
- Refactored, test, build and publish

## Version 0.1.0

- Release date: July 14, 2020 | refactored from prior-module written in nodejs
- Send Email / Messages Package


test/testData.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-15
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: test data for mc-mail testing
 */

export const verifyData = {
    email    : 'abbeymart@hotmail.com',
    username : 'abbeymart',
    loginName: 'abbeymart',
    msgFrom  : 'support@mconnect.biz',
    token    : 'testing-testing-testing',
    fullName : 'Abbey City',
    reqUrl   : 'mconnect.biz',
    type     : 'verify',
};

export const verifyRes = {
    code   : 'success',
    message: 'Verification email sent to your registered email, check and verify your email before login.',
}
export const verifyError = {
    code   : 'emailError',
    message: 'Error sending verification email, contact system-admin',
}

export const verifyUserRequestData = {
    email    : 'abbeymart@hotmail.com',
    username : 'abbeymart',
    loginName: 'abbeymart',
    msgFrom  : 'support@mconnect.biz',
    token    : 'testing-testing-testing',
    fullName : 'Abbey City',
    reqUrl   : 'mconnect.biz',
    type     : 'verifyUserRequest',
};
export const verifyUserReqRes = {
    code   : 'success',
    message: 'Verification email sent to your registered email, check and verify your email before login.',
}
export const verifyUserReqError = {
    code   : 'emailError',
    message: 'Error sending verification email, contact system-admin',
}

export const resetPasswordRequestData = {
    email    : 'abbeymart@hotmail.com',
    username : 'abbeymart',
    loginName: 'abbeymart',
    msgFrom  : 'support@mconnect.biz',
    token    : 'testing-testing-testing',
    fullName : 'Abbey City',
    reqUrl   : 'mconnect.biz',
    type     : 'resetPasswordRequest',
};
export const resetPassRes = {
    code   : 'success',
    message: 'Reset password request email sent to your registered email address.',
}
export const resetPassError = {
    code   : 'emailError',
    message: 'Error sending reset password request email, contact system-admin:',
}

export const setPasswordData = {
    email    : 'abbeymart@hotmail.com',
    username : 'abbeymart',
    loginName: 'abbeymart',
    msgFrom  : 'support@mconnect.biz',
    token    : 'testing-testing-testing',
    fullName : 'Abbey City',
    reqUrl   : 'mconnect.biz',
    type     : 'setPassword',
};
export const setPassRes = {
    code   : 'success',
    message: 'Password reset email sent to your registered email address',
}
export const setPassError = {
    code   : 'emailError',
    message: 'Error sending password reset email, contact system-admin:',
}

export const getLoginNameData = {
    email    : 'abbeymart@hotmail.com',
    username : 'abbeymart',
    loginName: 'abbeymart',
    msgFrom  : 'support@mconnect.biz',
    token    : 'testing-testing-testing',
    fullName : 'Abbey City',
    reqUrl   : 'mconnect.biz',
    type     : 'getLoginName',
};
export const getLoginNameRes = {
    code   : 'success',
    message: 'Login-name email sent to your registered email address.',
}
export const getLoginNameError = {
    code   : 'emailError',
    message: 'Error sending login-name email, contact system-admin:',
}

export const changePasswordData = {
    email    : 'abbeymart@hotmail.com',
    username : 'abbeymart',
    loginName: 'abbeymart',
    msgFrom  : 'support@mconnect.biz',
    token    : 'testing-testing-testing',
    fullName : 'Abbey City',
    reqUrl   : 'mconnect.biz',
    type     : 'changePassword',
};
export const changePassRes = {
    code   : 'success',
    message: 'Password reset email sent to your registered email address',
}
export const changePassError = {
    code   : 'emailError',
    message: 'Error sending password reset email, contact system-admin:',
}

export const defaultError = {
    code   : 'emailError',
    message: 'Error sending email: missing the required email parameters (type, title etc.)',
}


test/email.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-15
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: testing of the mc-mail package
 */

import { assertEquals, mcTest, postTestResult } from '@mconnect/mctest';
import { newEmail } from "../src";

import { EmailConfigType, EmailRequestType, EmailTemplateType } from "../src";
import { contactInfo, emailConfig, toEmailAddress } from "./config/emailConfig";
import { verifySubject, verifyContentText, verifyContentHtml } from "./templates";

const serverConfig: EmailConfigType = {
    username : emailConfig.username,
    password : emailConfig.password,
    serverUrl: emailConfig.serverUrl,
    port     : emailConfig.port,
    msgFrom  : emailConfig.msgFrom,
};

const requestInfo: EmailRequestType = {
    fromEmail   : emailConfig.msgFrom,
    toEmail     : toEmailAddress,
    successMessage: "success",
    requestName: "Guest User",
    templateData: {
        name       : "Abbey",
        verifyLink : "http://localhost:8080/verify",
        contactInfo: contactInfo,
        taskSubject: "Verify Registered Account",
    },
};

const emailTemp: EmailTemplateType = {
    subject: verifySubject,
    text   : verifyContentText,
    html   : verifyContentHtml,
};

const verifyRes = {
    code: "success",
    resMessage: "Requested Email sent to your registered email",
    message: "success",
};
const verifyError = {
    code: "emailError",
    message: "",
};

(async () => {
    await mcTest({
        name    : "should send and return valid response for verification",
        testFunc: async () => {
            const mail = newEmail(serverConfig)
            const mailRes = await mail.sendEmail(requestInfo, emailTemp);
            console.log("mail-response: ", mailRes)
            if (mailRes.code === 'success') {
                assertEquals(mailRes.code, verifyRes.code);
                assertEquals(mailRes.message.includes(verifyRes.message), true);
            } else {
                assertEquals(mailRes.code, verifyError.code);
            }
        },
    });

    await postTestResult();

})();


test/templates/index.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-08-26 | @Updated: 2020-07-14
 * @Company: mConnect.biz | @License: MIT
 * @Description: email templates
 */

export * from "./verify"
export * from "./changePass";
export * from "./resetPass";
export * from "./setPass";
export * from "./loginName";
export * from "./changeRequest";
export * from "./changeTask"


src/types.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-14
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-mail: types
 */

import { Readable } from "nodemailer/lib/xoauth2";
import { AttachmentLike } from "nodemailer/lib/mailer";

export interface EmailConfigType {
    username: string;
    password: string;
    port: number;
    serverUrl: string;
    msgFrom: string;
}

export interface EmailPropsType {
    [key: string]: string | number | Date;
}

export type EmailSubjectFunc = (props: EmailPropsType) => string;

export type EmailFunc = (props: EmailPropsType) => string | Buffer | Readable | AttachmentLike | undefined;

export interface TemplateDataType {
    [key: string]: string | number | Date;
}

export interface EmailRequestType {
    toEmail: Array<string>;
    requestName?: string;
    successMessage?: string;
    templateData: TemplateDataType;
    fromEmail?: string;
}

export interface EmailTemplateType {
    subject: EmailSubjectFunc;
    text: EmailFunc;
    html?: EmailFunc;
}

export interface MessageObject {
    [key: string]: string;
}


src/helper.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-12-15
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-crud helper functions
 */

import { getResMessage, ResponseMessage } from "@mconnect/mcresponse";
import { EmailConfigType } from "./types";

export function isEmptyObject(val: object): boolean {
    return !(Object.keys(val).length > 0 && Object.values(val).length > 0);
}

export function validateConfig(config: EmailConfigType): ResponseMessage {
    if (isEmptyObject(config) || config.serverUrl === "" || config.port === 0 || config.username === "" || config.password === "" || config.msgFrom === "") {
        return getResMessage("paramsError", {
            message: "email-configuration: user, password, server, port and message-from are required",
            value  : null,
        })
    } else {
        return getResMessage("success", {
            message: "email-configuration successfully validated",
            value  : null,
        })
    }
}

export const mcMessages = {
    loginError           : "Login audit error",
    logoutError          : "Logout audit error",
    authErrorMessage     : "Authorization error",
    confirmCashPayment   : "Confirm cash-payment completed / received, to complete order - It cannot be undone",
    confirmInvoicePayment: "Confirm invoice-payment (pay-on-delivery, cheque, wire transfer), to complete order",
    confirmDelete        : "Do you really want to permanently delete this record? It cannot be undone",
    confirmMsgDelete     : "Do you really want to delete this message?",
    confirmMsgUndelete   : "Do you really want to un-delete/recover this message?",
    itemMissing          : "Item not selected or missing, please try again",
    updateDenied         : "You are not authorized / permitted to update the record",
    docDuplicate         : "Record exist, record must be unique",
    docExists            : "No record created/changed, record with the same name exists",
    insertSuccess        : "New record saved successfully",
    updateStatus         : "Record updated successfully",
    validateInputs       : "Required information must be provided in the required format",
    subItems             : "Record contains sub-items and may not be removed",
    recordDeleted        : "Record removed successfully",
    removeDenied         : "You are not authorized / permitted to remove this record / item",
    noRecord             : "Record does not exist",
    langCodeFormat       : "Language code should be of the format en-US",
    nameFormat           : "Name should contains letters, words and spaces only",
    postalCodeFormat     : "Postal code contains letters, numbers and spaces only",
    numberFormat         : "Only numbers (0-9) are permitted",
    floatFormat          : "Currency / Float: only numbers (0-9) and . or , are permitted",
    phoneFormat          : "Only numbers (0-9), -, (, and ) are permitted",
    bnFormat             : "Only numbers (0-9) and - are permitted",
    titleFormat          : "Title should contains words/letters, spaces and hyphens only",
    descFormat           : "Description contains unacceptable inputs... include permitted formats only",
    codeFormat           : "Code should contain letters and numbers only",
    currencyFormat       : "Currency format is required: $, #, CA$, US$...",
    pathFormat           : "Path (tag) should contain letters and numbers only",
    urlFormat            : "URL should contain alphanumeric only",
    usernameFormat       : "Username should contains character/letters only",
    langRequired         : "Language is required",
    nameRequired         : "Name ( is required - cannot be empty",
    descRequired         : "Description is required - cannot be empty",
    catRequired          : "Category is required - cannot be empty",
    codeRequired         : "Code is required - cannot be empty",
    infoRequired         : "Information item is required",
    typeRequired         : "Type is required",
    groupRequired        : "Group is required",
    numberRequired       : "Number (0-9) is required - cannot be empty",
    titleRequired        : "Title is required - cannot be empty",
    emailRequired        : "Valid Email is required",
    emailMessage         : "Valid Email should be in the correct format",
    loginRequired        : "Valid Email or Username is required",
    passwordRequired     : "Password is required",
    passwordMessage      : "Password should include at least an upper-case and special character(@, #, $...)",
    passwordConfirm      : "Password and confirm-Password must be the same",
    confirmPassRequired  : "Confirmed password is required",
    stdCodeRequired      : "Standard code name may contains letters, numbers, spaces and hyphens only",
    contentErrorMessage  : "Content information should include valid title, category type and language",
    updateErrorMessage   : "Unable to update record - try again or contact system administrator",
    insertErrorMessage   : "Unable to create new record  - try again or contact system administrator",
    removeErrorMessage   : "Unable to remove record  - try again or contact system administrator",
    searchErrorMessage   : "Unable to retrieve record(s)  - try again or contact system administrator",
    paramErrorMessage    : "Error validating the inputs - information required in the acceptable formats",
    updateNotAuthorized  : "You are not authorized / permitted to update the record",
    noUpdate             : "Update not completed - current record ID is unknown or not provided",
    noInfo               : "Information not available... no data/records available",
    accessKeyMissing     : "Access-key missing or expired. Please login again.",
    orderMsgSubject      : "Order Details",
    cancelMsgSubject     : "Order / Order-Item(s) Cancelled",
    returnMsgSubject     : "Order-Item(s) Returned",
    payMsgSubject        : "Payment Made",
    shipMsgSubject       : "Order/Order-Items Shipped",
    refundMsgSubject     : "Order Payment Refund",
    getRequired          : "GET method is acceptable/required for this action/task",
    postRequired         : "POST method is acceptable/required for this action/task",
    getPostRequired      : "GET or POST method is acceptable/required for this action/task",
    cancelTypeOrder      : "Order - All Items",
    cancelTypeItem       : "Item(s) - Part of Order",
    isStringAlpha        : "Parameter should contains alphanumeric (a-z, A-Z, 0-9, -, _) only",
    isObject             : "Parameter should be object ({})",
    isArray              : "Parameter should be array ([])",
    usernameExists       : "Username exists, a unique username is required",
    emailExists          : "Email exists, a unique email is required",
    noPageInfo           : "Information is not available for this page",
};


src/sendAuthEmail.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-06-03 | @Updated: 2020-07-15
 * @Company: mConnect.biz | @License: MIT
 * @Description: mc-sendmail function
 */

import { isEmptyObject, validateEmailConfig } from "./helper";
import * as nodemailer from "nodemailer";
import { getResMessage, ResponseMessage } from "@mconnect/mcresponse";
import { EmailRequestType, EmailTemplateType, EmailConfigType } from "./types";
import Mail from "nodemailer/lib/mailer";

class Email {
    protected emailUser: string;
    protected emailPassword: string;
    protected emailPort: number;
    protected emailServer: string;
    protected request: EmailRequestType;
    protected template: EmailTemplateType
    protected transporter: Mail | undefined
    protected templateData: object

    constructor(config: EmailConfigType, request: EmailRequestType, template: EmailTemplateType, templateData: object) {
        this.emailUser = config.emailUser
        this.emailPassword = config.emailPassword
        this.emailPort = config.emailPort
        this.emailServer = config.emailServer
        this.request = request
        this.template = template
        this.templateData = {}
    }

    validateEmailConfig() {
        // validate email configuration
        const valRes = validateEmailConfig({
            emailUser    : this.emailUser,
            emailPassword: this.emailPassword,
            emailServer  : this.emailServer,
            emailPort    : this.emailPort,
        })
        if (valRes.code !== "success") {
            return valRes
        }
    }

    validateEmailRequestParams() {
        // required email parameters validation
        if ((typeof this.request !== "object") || (typeof this.request === "object" && isEmptyObject(this.request))) {
            return getResMessage("paramError", {
                message: "Invalid email parameters/inputs provided. Could not send email/message.",
            });
        }
    }

    activateMailServer() {
        this.transporter = nodemailer.createTransport({
            host          : this.emailServer,
            port          : this.emailPort,
            secure        : true, // use TLS
            auth          : {
                user: this.emailUser,
                pass: this.emailPassword,
            },
            tls           : {
                // do not fail on invalid certs
                rejectUnauthorized: false
            },
            pool          : true,
            maxMessages   : 200,
            maxConnections: 20,
        });
    }

    validateMailServer() {
        if (!this.transporter) {
            return getResMessage("emailError", {
                message: "Email Server Setup Error. Valid transporter is required to send email/message.",
            });
        }
    }

    async sendEmail(): Promise<ResponseMessage> {
        try {
            this.validateEmailConfig()
            this.validateEmailRequestParams()
            this.activateMailServer()
            this.validateMailServer()

            // send email
            const result = await this.transporter?.sendMail({
                from   : `${this.request.fromEmail}`, // sender address
                to     : `${this.request.toEmail}`, // list of receivers
                subject: this.template.subject(this.templateData), // Subject line
                text   : this.template.text(this.templateData), // plain text body
                html   : this.template.html ? this.template.html(this.templateData) : ""  // html body
            });
            return getResMessage("success", {
                message: "Verification email sent to your registered email, check and verify your email before login.",
                value  : result,
            });
        } catch (e) {
            return getResMessage("emailError", {
                message: `Error sending email: ${e.message}`,
            });
        }
    }
}

function newEmail(config: EmailConfigType, request: EmailRequestType, template: EmailTemplateType, templateData: object) {
    return new Email(config, request, template, templateData)
}

export { Email, newEmail }


src/sendEmail.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-06-03 | @Updated: 2020-07-15
 * @Company: mConnect.biz | @License: MIT
 * @Description: mc-sendmail function
 */

import { isEmptyObject, validateConfig } from "./helper";
import * as nodemailer from "nodemailer";
import { getResMessage, ResponseMessage } from "@mconnect/mcresponse";
import { EmailRequestType, EmailTemplateType, EmailConfigType, TemplateDataType } from "./types";
import Mail = require("nodemailer/lib/mailer");

class Email {
    protected emailUser: string;
    protected emailPassword: string;
    protected emailPort: number;
    protected serverUrl: string;
    protected msgFrom: string;
    protected request: EmailRequestType;
    protected template: EmailTemplateType;
    protected transporter: Mail | undefined;
    protected templateData: TemplateDataType;

    constructor(config: EmailConfigType) {
        this.emailUser = config.username
        this.emailPassword = config.password
        this.emailPort = config.port
        this.serverUrl = config.serverUrl
        this.msgFrom = config.msgFrom
        this.request = {fromEmail: "", toEmail: [], successMessage: "", requestName: "", templateData: {}}
        this.template = {text: ({}) => "", subject: ({}) => ""}
        this.templateData = {}
    }

    validateEmailConfig() {
        // validate email-server configuration
        const valRes = validateConfig({
            username : this.emailUser,
            password : this.emailPassword,
            serverUrl: this.serverUrl,
            port     : this.emailPort,
            msgFrom  : this.msgFrom,
        })
        if (valRes.code !== "success") {
            return valRes
        }
    }

    validateEmailRequestParams() {
        // required email parameters validation
        if (!this.request.toEmail) {
            return getResMessage("paramError", {
                message: "Request fromEmail and toEmail are required.",
            });
        }
        if ((typeof this.request.templateData !== "object") || (typeof this.request.templateData === "object" && isEmptyObject(this.request.templateData))) {
            return getResMessage("paramError", {
                message: "Email content-data is required to compose the email-message from the email-templates.",
            });
        }
    }

    activateMailServer() {
        // environment:
        const sysEnv = process.env.MC_ACCESS_ENV || "development";
        this.transporter = nodemailer.createTransport({
            host          : this.serverUrl,
            port          : this.emailPort,
            secure        : sysEnv !== "development", // use TLS for TEST/PROD deployment
            auth          : {
                user: this.emailUser,
                pass: this.emailPassword,
            },
            tls           : {
                // do not fail on invalid certs
                rejectUnauthorized: false
            },
            pool          : true,
            maxMessages   : 200,
            maxConnections: 20,
        });
    }

    validateMailServer() {
        if (!this.transporter) {
            return getResMessage("emailError", {
                message: "Email Server Setup Error. Valid transporter is required to send email/message.",
            });
        }
    }

    async sendEmail(request: EmailRequestType, template: EmailTemplateType): Promise<ResponseMessage> {
        try {
            this.request = request
            this.template = template
            this.templateData = request.templateData
            // validate send-email parameters / inputs
            this.validateEmailConfig()
            this.validateEmailRequestParams()
            this.validateMailServer()

            // activate email-server
            this.activateMailServer()

            // send email
            const result = await this.transporter?.sendMail({
                from   : `${this.request.fromEmail}`, // sender address
                to     : this.request.toEmail, // list of receivers
                subject: this.template.subject(this.templateData), // Subject line
                text   : this.template.text(this.templateData), // plain text body
                html   : this.template.html ? this.template.html(this.templateData) : ""  // html body
            });
            return getResMessage("success", {
                message: this.request.requestName ? (this.request.successMessage ? `${this.request.requestName}: ${this.request.successMessage}` :
                    `${this.request.requestName}: Requested Email sent to your registered email. Check you email for details.`) :
                    "Requested Email sent to your registered email. Check you email for details.",
                value  : result,
            });
        } catch (e) {
            return getResMessage("emailError", {
                message: `Error sending email: ${e.message}`,
            });
        }
    }
}

function newEmail(config: EmailConfigType) {
    return new Email(config)
}

export { Email, newEmail }
